plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java

```diff
@@ -213,24 +214,53 @@ private void setupProjectVersionDataBinding(DataBindingContext context) {
+    ISWTObservableValue promoteButton =
+        WidgetProperties.selection().observe(autoPromoteButton);
+    final ISWTObservableValue stopPreviousVersion =
         WidgetProperties.selection().observe(stopPreviousVersionButton);
-    ISWTObservableValue stopPreviousVersionEnablement =
+    final ISWTObservableValue stopPreviousVersionEnablement =
         WidgetProperties.enabled().observe(stopPreviousVersionButton);
 
-    // use an intermediary value to control the enabled state of stopPreviousVersionButton
-    // based on the promote checkbox's state
-    WritableValue enablement = new WritableValue();
-    context.bindValue(promoteButton, enablement);
-    context.bindValue(stopPreviousVersionEnablement, enablement);
+    context.bindValue(stopPreviousVersionEnablement, promoteButton);
 
     IObservableValue promoteModel = PojoProperties.value("autoPromote").observe(model);
     IObservableValue stopPreviousVersionModel =
         PojoProperties.value("stopPreviousVersion").observe(model);
 
     context.bindValue(promoteButton, promoteModel);
-    context.bindValue(stopPreviousVersion, stopPreviousVersionModel);
+
+    // Intermediary model necessary for "Restore Defaults" to work.
+    final IObservableValue currentStopPreviousVersionChoice = new WritableValue();
+    context.bindValue(currentStopPreviousVersionChoice, stopPreviousVersionModel);
+
+    // One-way update: button selection <-- latest user choice
+    // Update the button (to match the user choice), if enabled; if not, force unchecking.
+    context.bindValue(stopPreviousVersion, new ComputedValue() {
+      @Override
+      protected Object calculate() {
+        boolean buttonEnabled = (boolean) stopPreviousVersionEnablement.getValue();
+        boolean currentValue = (boolean) currentStopPreviousVersionChoice.getValue();
+        if (!buttonEnabled) {
+          return Boolean.FALSE;  // Force unchecking the stop previous button if it is disabled.
+        }
+        return currentValue;  // Otherwise, check the button according to the latest user choice.
+      }
+    }, new UpdateValueStrategy(UpdateValueStrategy.POLICY_NEVER), null);
+
+    // One-way update: button selection --> latest user choice
+    // Update the user choice (to match the button selection), only when the button is enabled.
+    context.bindValue(new ComputedValue() {
+      @Override
+      protected Object calculate() {
+        boolean buttonEnabled = (boolean) stopPreviousVersionEnablement.getValue();
+        boolean buttonValue = (boolean) stopPreviousVersion.getValue();
+        boolean currentValue = (boolean) currentStopPreviousVersionChoice.getValue();
+        if (buttonEnabled) {
+          return buttonValue;  // Remember the button state as the latest choice if it is enabled.
+        }
+        return currentValue;  // Otherwise, retain the latest (current) user choice.
+      }
+    }, stopPreviousVersionModel, null, new UpdateValueStrategy(UpdateValueStrategy.POLICY_NEVER));
```

Comment:
Ugh.  This is the situation for Neon's _SideEffect_.  I wonder if it's worth rolling our own for Mars.